### PR TITLE
New azure_rag pipeline

### PR DESCRIPTION
**Pull request by richmondsin**
This is the new pipeline for data processing for the azure indexing.

Files edited:
catalog.yml
- Added filtered_data_rag, processed_data_rag and json_data_rag.

./base/parameters_azure_rag.yml
- Added the parameters for the functions in nodes.py.

./pipelines/azure_rag/nodes.py
- Added the 3 functions: filtering of articles by their "removal_type", processing HTML tables by the LLM, and also the partition of the data frame into the respective "content" and "table" json files.

./pipelines/azure_rag/pipeline.py
- Added the inputs and outputs for the respective functions: merged_data -> filtered_data_rag, filtered_data_rag -> processed_data_rag, processed_data_rag -> json_data_rag.

New folders but unedited:
./tests/azure_rag

### CODE REVIEW

**Comment by Wilsven** [comment]
Hi, could you please elaborate on your PR? It is best practice, as mentioned before.

**Comment by Wilsven** [comment]
> Hi, could you please elaborate on your PR? It is best practice, as mentioned before.

Your branch is also not up-to-date with the main branch. Have you pulled in the latest changes to your branch before making a PR?

**Review by Wilsven — CHANGES_REQUESTED** [review]
- Update in detail changes in this PR
- Branch should be up-to-date

File: content-optimization/requirements.txt [hunk 1]
```diff
@@ -1,6 +1,9 @@
 alive_progress==3.1.5
+azure-cli
```
Comment (Wilsven) [inline content-optimization/requirements.txt:2]: Can you specify the version?

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+def ask(html_content: str) -> str:
+    azure_credential = DefaultAzureCredential()
+    token_provider = get_bearer_token_provider(
+        azure_credential, "https://cognitiveservices.azure.com/.default"
+    )
+
+    openai_client = AzureOpenAI(
+        api_version="2024-03-01-preview",
+        azure_endpoint="https://cog-zicuh2lqrbb2s.openai.azure.com",
```
Comment (Wilsven) [inline content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py:13]: This endpoint should not be committed to version control. Please read up on how to deal with credentials using Kedro.

File: content-optimization/src/content_optimization/pipelines/azure_rag/nodes.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    # Remove the rows with 'NaN' from 'remove_type' column
+    filtered_data_rag = filtered_data_rag[filtered_data_rag["remove_type"] != "NaN"]
+    # Remove 'Multilingual' from 'remove_type' column
+    filtered_data_rag = filtered_data_rag[
+        filtered_data_rag["remove_type"] != "Multilingual"
+    ]
+
+    # Remove the duplicated articles with specific 'id' values
+    filtered_data_rag = filtered_data_rag[
+        ~filtered_data_rag["id"].isin([1445629, 1443608, 1435183, 1435335, 1434652])
```
Comment (Wilsven) [inline content-optimization/src/content_optimization/pipelines/azure_rag/nodes.py:30]: Please do not hardcode. Use the configuration file.

**Review by Wilsven — CHANGES_REQUESTED** [review]

File: content-optimization/src/content_optimization/pipelines/azure_rag/nodes.py (hunk 3 above)
Comment (richmondsin) [reply content-optimization/src/content_optimization/pipelines/azure_rag/nodes.py]: Done.

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py (hunk 2 above)
Comment (richmondsin) [reply content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py]: Done.

File: content-optimization/requirements.txt (hunk 1 above)
Comment (richmondsin) [reply content-optimization/requirements.txt]: Done.

**Comment by richmondsin** [comment]
> Hi, could you please elaborate on your PR? It is best practice, as mentioned before.

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            "role": "system",
+            "content": "You are an AI assistant specialized in extracting structured content from HTML.",
+        },
+        {"role": "user", "content": prompt.format(html_content=html_content)},
+    ]
+
+    response = openai_client.chat.completions.create(
+        messages=query_messages,
+        model="chat",
```
Comment (Wilsven) [inline content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py:49]: Can you put the deployment name in the configuration file or credential file as well? Either one will do.

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        temperature=0.0,
+        max_tokens=2000,
```
Comment (Wilsven) [inline content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py:51]: All these parameters can go into the configuration file. You never know when we might want to tweak these.

**Review by Wilsven — CHANGES_REQUESTED** [review]

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py (hunk 4 above)
Comment (richmondsin) [reply content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py]: Done.

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py (hunk 5 above)
Comment (richmondsin) [reply content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py:52]: But when they tweak this, they will look into this script. And this is a separate python file that I import into the nodes.py so the parameters.yml is not tied to this.

Comment (Wilsven) [reply content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py:52]: You are making the assumption they do refer to your script. Your script in this pipeline shouldn't be looked at when we are running iterations on the pipeline. If we want to run it with different parameters, it should be as simple as configuring your configuration file. 

I do not get your point on 

> And this is a separate python file that I import into the nodes.py so the parameters.yml is not tied to this.

I do not get your point on this. Even if it is imported from a `utils.py`, these are just functions where arguments to one function can be the arguments of another function. Your node function can simply take as parameters these arguments.

**Review by Wilsven — CHANGES_REQUESTED** [review]
Resolve remaining comments.

File: content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py (hunk 5 above)
Comment (richmondsin) [reply content-optimization/src/content_optimization/pipelines/azure_rag/llm_extraction.py]: Done.

**Comment by Wilsven** [comment]
@richmondsin Could you please update the `README.md` in the [File Structure section](https://github.com/Synapxe-DNA/healthhub-content-optimization/tree/main?tab=readme-ov-file#file-structure) as to what the data you committed under the `01_raw` is?

**Pull request merged**